add initial functional requirements [draft] (#4)
haveno-dex/haveno-design: functional_requirements.md

# Haveno Functional Requirements

This document maintains functional requirements for the Haveno user interface (work in progress).

### Startup
- On first startup, the user must configure a password.

> - On first startup, the user can use a local node or configure a remote node (url, username, password).

Reviewer: ```suggestion
- On first startup, the user is asked if they want to use a local node or a remote one. If they choose 'local' Haveno will look for a local node. If they choose 'remote', they are prompted to another choice: 'Automatically detect' node or 'specify node' (defaults to Haveno public node) where they specify url + port, username (not mandatory), password (not mandatory)
```

Reviewer 2: When selecting the option to use a local node, it sets up a node on the user's machine, right? 

In the case of using a remote node: how can we automatically detect which remote node the user wants to use? What is this based on?

Reviewer: > When selecting the option to use a local node, it sets up a node on the user's machine, right?

No. It will look for a local node that is already running locally.

> how can we automatically detect which remote node the user wants to use? What is this based on?

The Monero GUI and Monerujo support scanning the network looking for valid remote nodes. We are considering implementing Monerujo's version of this feature (which is already in java) in the mvp, but it's not certain. What's certain is that there will be a hard coded node address that people will be able to change if they want to rely on a specific remote node.

Reviewer 2: Got it.

- On startup, the application presents a splash screen with progress and description of startup state.

> ### Wallet

Reviewer 2: Should a user be able to create a new wallet?

Author: The application has exactly one "main wallet" which is auto-generated on first startup and doesn't change, so I don't see a use case for generating a new wallet.

Reviewer 2: Alright. Perhaps due to it being a monero wallet, the need is less high to create new wallets. I do know some Monero users doing it to be extra cautious.

Reviewer: @deaner0 I think you are referring to subaddress? That should be supported, yes.

- The user can deposit into their Haveno wallet.

> - The user can generate new deposit subaddresses for their Haveno wallet.

Reviewer: or a new one is generated for them?

Author: A new one should be generated automatically, and we can allow them to manually generate a new one.

- The user can view the time until the Haveno wallet balance unlocks (if locked).

### View Market
- For each trade asset, the user can view open offers to buy or sell XMR.
- For each trade asset, the user can view a graph showing market depth of open offers.

### Make or Take Offers
- The user can make an offer to buy or sell XMR for another asset.
- The user can take an offer to buy or sell XMR for another asset.
- The Haveno wallet must have sufficient balance in order to make or take an offer.
- If the Haveno wallet balance is insufficient to make or take an offer, the user is prompted to deposit funds to their Haveno wallet and wait 10 confirmations (~20 minutes).
- The user is prompted to confirm trade details before making or taking offer.
- Total funds needed to trade = trade amount + security deposit + trade fee.
- If the user is selling XMR, they can enter a new receiving payment account (which is saved) or select from previous receiving payment accounts.

### Open Trade
- The user can directly chat with their trade peer.
- The user can open a dispute for an open trade after an allotted time.

> - The user can directly chat with the arbitrator.

Reviewer: ```suggestion
- The user can directly chat with the arbitrator after a dispute is open.
```

- The user can accept or reject the arbitrator decision?
- The user is notified when their offer is taken.
- The user is notified when they should make payment or confirm received payment.

### View User's Offers and Trades
- The user can view and cancel their open offers.
- The user can view their open trades.
- The user can view their completed trades.

> - The user can view their failed trades.

Reviewer: ```suggestion
- The user can view their failed trades.
- The user can view disputed trades and their status
```

Author: Disputed trades are technically a subset of open trades. Don't know if we want a dedicated view for them or to incorporate a filter or flag on the existing view for open trades.
